### ISSUE TITLE
CUDA_POST_KERNEL_CHECK doesn't check errors after kernel execution

### ISSUE THREAD

**Issue by longjon** [BVLC/caffe#2501]
I'm not sure exactly what the intent was, since the `CUDA_POST_KERNEL_CHECK` macro dates from @Yangqing's aef20cdaa27087b4fbd20336cd0318ca92a4d260 commit, but the current comment and the macro name give the impression that this macro will check for any errors that occur during kernel execution.

I want to make a note here that that does not seem to be true (unless I am misunderstanding the CUDA API); `cudaPeekAtLastError` returns immediately after kernel launch, so launch errors should be reported, but errors that actually occur during kernel execution will be postponed for later.

(I believe there are other instances of this sort of thing, e.g., `CUDA_CHECK(cudaMemcpy(...))` will not check errors that occur during device-to-device copies, though maybe that is simply the cost of doing business; we should probably think more at some point about what should be synchronized when, and the implications for error checking...)

**Comment by Yangqing** [comment]
Yep, looks like a cudaDeviceSynchronize() is missing. Or if we are using non-default stream, a cudaStreamSynchronize().
